File: kumi/sdk/typescript/src/connection.ts
import { bootstrapProfile, isLanCode, isRelayToken, parseLanCode } from "./auth";
import { getKumiEnv } from "./runtime";

export interface ConnectionConfig {
  mode: "direct" | "relay";
  baseUrl: string;
  accessToken?: string;
}

export function httpToWs(url: string): string {
  if (url.startsWith("https://")) return "wss://" + url.slice("https://".length);
  if (url.startsWith("http://")) return "ws://" + url.slice("http://".length);
  return url;
}

export function relayEdgeWsUrl(config: ConnectionConfig): string {
  return httpToWs(config.baseUrl.replace(/\/+$/, "")) + "/ws/edge";
}

export async function resolveConnection(
  code: string | undefined,
  edgeName: string,
  env: Record<string, string | undefined> = getKumiEnv()
): Promise<ConnectionConfig> {
  const relayUrl = env.KUMI_RELAY_URL;
  const accessToken = env.KUMI_ACCESS_TOKEN;
  if (relayUrl && accessToken) {
    return {
      mode: "relay",
      baseUrl: relayUrl.replace(/\/+$/, ""),
      accessToken,
    };
  }

  const c = code ?? "";

  if (c.startsWith("ws://") || c.startsWith("wss://")) {
    return { mode: "direct", baseUrl: c };
  }

  if (isLanCode(c)) {
    const serverUrl = parseLanCode(c);
    const wsUrl = httpToWs(serverUrl.replace(/\/+$/, "")) + "/ws/edge";
    return { mode: "direct", baseUrl: wsUrl };
  }

  if (isRelayToken(c)) {
    const profile = await bootstrapProfile(c, "edge", edgeName);
    return {
      mode: "relay",
      baseUrl: profile.relayUrl,
      accessToken: profile.accessToken,
    };
  }

  if (c.startsWith("http://") || c.startsWith("https://")) {
    const wsUrl = httpToWs(c.replace(/\/+$/, "")) + "/ws/edge";
    return { mode: "direct", baseUrl: wsUrl };
  }

  return { mode: "direct", baseUrl: "ws://127.0.0.1:8000/ws/edge" };
}

File: kumi/sdk/csharp/Connection.cs
namespace Kumi;

/// <summary>
/// Connection resolution: determines WebSocket URL and access token
/// from environment variables and connection code.
/// </summary>
internal static class Connection
{
    public sealed class ConnectionConfig
    {
        public string Mode { get; }
        public string BaseUrl { get; }
        public string? AccessToken { get; }

        public ConnectionConfig(string mode, string baseUrl, string? accessToken)
        {
            Mode = mode;
            BaseUrl = baseUrl;
            AccessToken = accessToken;
        }

        public string RelayEdgeWsUrl() =>
            HttpToWs(BaseUrl.TrimEnd('/')) + "/ws/edge";
    }

    public static string HttpToWs(string url)
    {
        if (url.StartsWith("https://")) return "wss://" + url["https://".Length..];
        if (url.StartsWith("http://")) return "ws://" + url["http://".Length..];
        return url;
    }

    public static async Task<ConnectionConfig> ResolveConnection(string? code, string edgeName)
    {
        var relayUrl = Environment.GetEnvironmentVariable("KUMI_RELAY_URL");
        var accessToken = Environment.GetEnvironmentVariable("KUMI_ACCESS_TOKEN");
        if (!string.IsNullOrEmpty(relayUrl) && !string.IsNullOrEmpty(accessToken))
            return new ConnectionConfig("relay", relayUrl.TrimEnd('/'), accessToken);

        code ??= "";

        if (code.StartsWith("ws://") || code.StartsWith("wss://"))
            return new ConnectionConfig("direct", code, null);

        if (Auth.IsLanCode(code))
        {
            var serverUrl = Auth.ParseLanCode(code);
            var wsUrl = HttpToWs(serverUrl.TrimEnd('/')) + "/ws/edge";
            return new ConnectionConfig("direct", wsUrl, null);
        }

        if (Auth.IsRelayToken(code))
        {
            var profile = await Auth.BootstrapProfile(code, "edge", edgeName);
            return new ConnectionConfig("relay", profile.RelayUrl, profile.AccessToken);
        }

        if (code.StartsWith("http://") || code.StartsWith("https://"))
        {
            var wsUrl = HttpToWs(code.TrimEnd('/')) + "/ws/edge";
            return new ConnectionConfig("direct", wsUrl, null);
        }

        return new ConnectionConfig("direct", "ws://127.0.0.1:8000/ws/edge", null);
    }
}
